File: Frontend/services/dashboardService.ts
import { type DashboardMetrics } from "@/models/dashboard";

export async function getDashboardMetrics(): Promise<DashboardMetrics | null> {
  const py = process.env.NEXT_PUBLIC_PY_API_URL;
  if (!py) return null;

  try {
    const url = new URL(`${py}/api/dashboard/metrics`);
    const res = await fetch(url.toString(), { cache: "no-store" });
    if (!res.ok) return null;
    const data = await res.json();
    return {
      total_computers: Number(data.total_computers ?? 0),
      preventive_done_computers: Number(data.preventive_done_computers ?? 0),
      preventive_needed_computers: Number(data.preventive_needed_computers ?? 0),
      corrective_done_total: Number(data.corrective_done_total ?? 0),
      corrective_done_computers: Number(data.corrective_done_computers ?? 0),
      status_ok_computers: Number(data.status_ok_computers ?? 0),
      status_late_computers: Number(data.status_late_computers ?? 0),
      status_pending_computers: Number(data.status_pending_computers ?? 0),
      corrective_open_computers: Number(data.corrective_open_computers ?? 0)
    };
  } catch {
    return null;
  }
}

File: python-api/app/services/dashboard_service.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import distinct, func, or_
from sqlalchemy.orm import Session

from app.models import Computer, MaintenanceHistory
from app.schemas.schemas import DashboardMetrics


def get_dashboard_metrics(db: Session) -> DashboardMetrics:
    now = datetime.utcnow()

    total_computers = int(db.query(func.count(Computer.id)).scalar() or 0)
    preventive_done_computers = int(
        db.query(func.count(Computer.id))
        .filter(Computer.last_maintenance.isnot(None))
        .scalar()
        or 0
    )

    status_pending = int(
        db.query(func.count(Computer.id)).filter(Computer.next_maintenance.is_(None)).scalar() or 0
    )
    status_late = int(
        db.query(func.count(Computer.id))
        .filter(Computer.next_maintenance.isnot(None), Computer.next_maintenance < now)
        .scalar()
        or 0
    )
    status_ok = int(
        db.query(func.count(Computer.id))
        .filter(Computer.next_maintenance.isnot(None), Computer.next_maintenance >= now)
        .scalar()
        or 0
    )

    corrective_open = int(
        db.query(func.count(Computer.id))
        .filter(or_(Computer.next_maintenance.is_(None), Computer.next_maintenance < now))
        .scalar()
        or 0
    )

    corrective_done_total = int(
        db.query(func.count(MaintenanceHistory.id))
        .filter(MaintenanceHistory.maintenance_type == "Corretiva")
        .scalar()
        or 0
    )

    corrective_done_computers = int(
        db.query(func.count(distinct(MaintenanceHistory.computer_id)))
        .filter(MaintenanceHistory.maintenance_type == "Corretiva")
        .scalar()
        or 0
    )

    return DashboardMetrics(
        total_computers=total_computers,
        preventive_done_computers=preventive_done_computers,
        preventive_needed_computers=total_computers,
        corrective_done_total=corrective_done_total,
        corrective_done_computers=corrective_done_computers,
        status_ok_computers=status_ok,
        status_late_computers=status_late,
        status_pending_computers=status_pending,
        corrective_open_computers=corrective_open,
    )
